zerver/lib/streams.py

```diff
@@ -591,6 +595,49 @@ def access_stream_for_send_message(
     )
 
 
+def check_stream_access_based_on_can_create_topic_group(
+    user_profile: UserProfile, stream: Stream
+) -> None:
+    if is_cross_realm_bot_email(user_profile.delivery_email):
+        return
+
+    can_create_topic_group = stream.can_create_topic_group
+    if hasattr(can_create_topic_group, "named_user_group"):
+        if can_create_topic_group.named_user_group.name == SystemGroups.EVERYONE:
+            return
+
+        if can_create_topic_group.named_user_group.name == SystemGroups.NOBODY:
+            raise JsonableError(
+                _("You do not have permission to create new topics in this channel.")
+            )
+
+    if not user_has_permission_for_group_setting(
+        stream.can_create_topic_group_id,
+        user_profile,
+        Stream.stream_permission_group_settings["can_create_topic_group"],
+        direct_member_only=False,
+    ):
+        raise JsonableError(_("You do not have permission to create new topics in this channel."))
+
+
+def access_stream_to_create_new_topic(
+    user_profile: UserProfile, stream: Stream, topic_name: str
+) -> None:
+    assert stream.recipient_id is not None
+    topic_exists = messages_for_topic(
```

Comment:
do we require any handling for empty topics ?
other places have checks such as `("(no topic)", "")`